File: modules/vescape-core/ios/telemetry/DatabaseFileSwap.swift
import Foundation

/// Installs a staged SQLite file and restores the complete former file set on any failure.
internal func replacingDatabaseFiles<T>(
  source: URL,
  target: URL,
  install: (URL) throws -> T
) throws -> T {
  let fm = FileManager.default
  let suffixes = ["", "-wal", "-shm"]
  let rollbackDir = target.deletingLastPathComponent()
    .appendingPathComponent("\(target.lastPathComponent).rollback-\(UUID().uuidString)", isDirectory: true)
  try fm.createDirectory(at: rollbackDir, withIntermediateDirectories: true)
  var moved: [(original: URL, saved: URL)] = []
  do {
    for suffix in suffixes {
      let current = URL(fileURLWithPath: target.path + suffix)
      guard fm.fileExists(atPath: current.path) else { continue }
      let saved = rollbackDir.appendingPathComponent(target.lastPathComponent + suffix)
      try fm.moveItem(at: current, to: saved)
      moved.append((current, saved))
    }
    try fm.copyItem(at: source, to: target)
    let result = try install(target)
    try fm.removeItem(at: rollbackDir)
    return result
  } catch {
    let installError = error
    var rollbackErrors: [Error] = []
    for suffix in suffixes {
      let installed = URL(fileURLWithPath: target.path + suffix)
      if fm.fileExists(atPath: installed.path) {
        do { try fm.removeItem(at: installed) } catch { rollbackErrors.append(error) }
      }
    }
    for entry in moved {
      do { try fm.moveItem(at: entry.saved, to: entry.original) } catch { rollbackErrors.append(error) }
    }
    guard !rollbackErrors.isEmpty else {
      try fm.removeItem(at: rollbackDir)
      throw installError
    }
    let rollbackDetails = rollbackErrors.map { String(describing: $0) }.joined(separator: "; ")
    throw NSError(
      domain: "VescapeDatabaseSwap",
      code: 2,
      userInfo: [
        NSLocalizedDescriptionKey: "Database install failed and rollback was incomplete; preserved recovery files at \(rollbackDir.path)",
        NSUnderlyingErrorKey: installError,
        "rollbackErrors": rollbackDetails,
      ]
    )
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/telemetry/DatabaseFileSwap.kt
package expo.modules.vescapecore.telemetry

import java.io.File

/** Installs a staged SQLite file and restores the complete former file set on any failure. */
internal fun replaceDatabaseFiles(
  source: File,
  target: File,
  validateInstalled: (File) -> Unit,
) {
  val suffixes = listOf("", "-wal", "-shm")
  val rollbackDir = File(target.parentFile, "${target.name}.rollback")
  check(!rollbackDir.exists() || rollbackDir.deleteRecursively()) { "Could not clear database rollback directory" }
  check(rollbackDir.mkdirs()) { "Could not create database rollback directory" }
  val moved = mutableListOf<Pair<File, File>>()
  var cleanupRollback = false
  try {
    for (suffix in suffixes) {
      val current = File(target.path + suffix)
      if (!current.exists()) continue
      val saved = File(rollbackDir, target.name + suffix)
      check(current.renameTo(saved)) { "Could not preserve current database file ${current.name}" }
      moved += current to saved
    }
    source.copyTo(target, overwrite = false)
    validateInstalled(target)
    cleanupRollback = true
  } catch (restoreError: Exception) {
    val rollbackErrors = mutableListOf<Throwable>()
    for (suffix in suffixes) {
      val installed = File(target.path + suffix)
      if (installed.exists() && !installed.delete()) {
        rollbackErrors += IllegalStateException("Could not remove failed restored database file ${installed.name}")
      }
    }
    for ((original, saved) in moved) {
      if (!saved.renameTo(original)) {
        rollbackErrors += IllegalStateException("Could not roll back current database file ${original.name}")
      }
    }
    if (rollbackErrors.isEmpty()) {
      cleanupRollback = true
      throw restoreError
    }
    throw IllegalStateException(
      "Database install failed and rollback was incomplete; preserved recovery files at ${rollbackDir.path}",
      restoreError,
    ).also { combined -> rollbackErrors.forEach(combined::addSuppressed) }
  } finally {
    if (cleanupRollback && rollbackDir.exists()) {
      check(rollbackDir.deleteRecursively()) { "Could not clear database rollback directory" }
    }
  }
}
